src/System.Devices.Gpio/System.Devices.Gpio/UnixI2CDevice.cs

```diff
@@ -0,0 +1,349 @@
+
+using System.IO;
+using System.Runtime.InteropServices;
+
+namespace System.Devices.Gpio
+{
+    public class UnixI2cDevice : I2cDevice
+    {
+        #region Interop
+
+        private const string LibraryName = "libc";
+
+        [Flags]
+        private enum FileOpenFlags
+        {
+            O_RDONLY = 0x00,
+            O_NONBLOCK = 0x800,
+            O_RDWR = 0x02,
+            O_SYNC = 0x101000
+        }
+
+        [DllImport(LibraryName, SetLastError = true)]
+        private static extern int open([MarshalAs(UnmanagedType.LPStr)] string pathname, FileOpenFlags flags);
+
+        [DllImport(LibraryName, SetLastError = true)]
+        private static extern int close(int fd);
+
+        private enum I2cSettings : uint
+        {
+            /// <summary>Combined R/W transfer (one STOP only)</summary>
+            I2C_RDWR = 0x0707,
+            /// <summary>Smbus transfer</summary>
+            I2C_SMBUS = 0x0720,
+            /// <summary>Get the adapter functionality mask</summary>
+            I2C_FUNCS = 0x0705,
+            /// <summary>Use this slave address, even if it is already in use by a driver</summary>
+            I2C_SLAVE_FORCE = 0x0706
+        }
+
+        /// To determine what functionality is supported
+        [Flags]
+        private enum I2cFunctionalityFlags : ulong
+        {
+            I2C_FUNC_I2C = 0x00000001,
+            I2C_FUNC_SMBUS_BLOCK_DATA = 0x03000000
+        }
+
+        [Flags]
+        private enum I2cMessageFlags : ushort
+        {
+            /// <summary>Write data to slave</summary>
+            I2C_M_WR = 0x0000,
+            /// <summary>Read data from slave</summary>
+            I2C_M_RD = 0x0001
+        }
+
+        private unsafe struct i2c_msg
+        {
+            public ushort addr;
+            public I2cMessageFlags flags;
+            public ushort len;
+            public byte* buf;
+        };
+
+        /// <summary>Used in the <see cref="I2C_RDWR"/> <see cref="ioctl"/> call</summary>
+        private unsafe struct i2c_rdwr_ioctl_data
+        {
+            public i2c_msg* msgs;
+            public uint nmsgs;
+        };
+
+        private enum SmbusMessageFlags : byte
+        {
+            /// <summary>Write data to slave</summary>
+            I2C_SMBUS_WRITE = 0,
+            /// <summary>Read data from slave</summary>
+            I2C_SMBUS_READ = 1
+        }
+
+        private enum SmbusTransactionType : uint
+        {
+            I2C_SMBUS_BLOCK_DATA = 5
+        }
+
+        private unsafe struct i2c_smbus_ioctl_data
+        {
+            public SmbusMessageFlags read_write;
+            public SmbusTransactionType size;
+            public byte command;
+            public byte* data;
+        }
+
+        /// <summary>As specified in SMBus standard</summary>
+        private const int I2C_SMBUS_BLOCK_MAX = 32;
+
+        [StructLayout(LayoutKind.Explicit)]
+        private unsafe struct i2c_smbus_data
```

Comment:
which native header file did these struct definitions come from?  Could you add a comment pointing to the source so that future maintainers can find it more easily?